File: videocode/input/shape/Circle.py
#!/usr/bin/env python3

from __future__ import annotations

import math

from videocode.input.shape.Polygon import Polygon, paint
from videocode.utils.decorators import prop
from videocode.ty import *
from videocode.constants import *


class Circle(Polygon):
    """
    A circle approximated by quadratic bezier segments — a `Polygon` whose
    control points are generated directly (raw contour), so it gets the full
    polygon pipeline: geometry cache, gradients, zIndex, effects.
    """

    def __init__(
        self,
        radius: wufloat = 1,
        fillColor: paint = RED_A,
        strokeColor: rgba = WHITE,
        strokeWidth: wufloat = 0.05,
    ):
        self.radius = radius
        super().__init__(
            vertices=self.generateVertices(),
            fillColor=fillColor,
            strokeColor=strokeColor,
            strokeWidth=strokeWidth,
        )

    def generateRawContours(self) -> list[list[point]]:
        # These points are already final bezier anchor/handle pairs, so this
        # bypasses Polygon's default corner-rounding (which expects one plain
        # vertex per corner and emits 4 bezier points per corner — it would
        # mangle an already-bezier contour like this one).
        #
        # Manim's formula: anchors on the circle; handles at the mid-arc angle,
        # scaled outward by 1/cos(segAngle/2) (the tangent-intersection point).
        # Emitted clockwise (negative sin) so the pixel-space winding after the
        # renderer's y-negation matches the other polygons.
        N = 16
        segAngle = 2 * math.pi / N
        cosHalf = math.cos(segAngle / 2)
        r = self.radius

        pts: list[point] = []
        for i in range(N):
            aAngle = i * segAngle
            hAngle = aAngle + segAngle / 2
            pts.append((r * math.cos(aAngle), -r * math.sin(aAngle)))
            pts.append((r / cosHalf * math.cos(hAngle), -r / cosHalf * math.sin(hAngle)))
        return [pts]

    def generateVertices(self) -> list[point]:
        # Required by Polygon (used for width/height/contains-point checks).
        # Circle has no separate "corner vertex" representation, so this just
        # reuses the same raw contour points as generateRawContours.
        return list(self.generateRawContours()[0])

    @prop(onSet=Polygon.updatePoints)
    def radius() -> wufloat: ...


class Dot(Circle):
    def __init__(
        self,
        radius: float = 0.0375,
        fillColor: paint = RED_A,
    ):
        super().__init__(
            radius=radius,
            fillColor=fillColor,
            strokeColor=TRANSPARENT,
            strokeWidth=0,
        )

File: src/input/shape/Circle.cpp
/*
** EPITECH PROJECT, 2025
** video-code
** File description:
** Circle
*/

#include "input/shape/Circle.hpp"

#include <cmath>

Circle::Circle(json::object_t&& args)
    : BezierPath(std::move(args))
{
}

// Approximate a circle with N quadratic bezier segments using Manim's formula:
// anchors lie on the circle; handles are the mid-arc point scaled outward by
// 1/cos(segAngle/2), which equals the tangent-intersection distance.
void Circle::buildPath(const json::object_t& args)
{
    float radius = args.at("radius").get<float>() * config::worldToPixelRatio;
    _strokeWidth = args.at("strokeWidth").get<float>() * config::worldToPixelRatio;
    parseColorOrGradient(args, "fillColor",   _fillColor,   _fillStops,   _fillGradType,   _fillGradientAngle);
    parseColorOrGradient(args, "strokeColor", _strokeColor, _strokeStops, _strokeGradType, _strokeGradientAngle);
    _closed = true;

    const int   N = 16;
    const float segAngle = 2.f * static_cast<float>(M_PI) / N; // 45°
    const float cosHalf = std::cos(segAngle / 2.f);            // cos(22.5°) ≈ 0.9239

    _points.clear();
    _points.reserve(N * 2);

    for (int i = 0; i < N; ++i) {
        float aAngle = i * segAngle;
        float hAngle = aAngle + segAngle / 2.f;
        _points.push_back(radius * cv::Vec2f{std::cos(aAngle), std::sin(aAngle)}); // not reversed
        _points.push_back((radius / cosHalf) * cv::Vec2f{std::cos(hAngle), std::sin(hAngle)});
    }
}
